File: Coding/golang/timerqueue/timerqueue.go
package main

import (
	"container/heap"
	"fmt"
	"sync"
	"time"
)

// TimerTask states.
const (
	taskNew       = 1
	taskScheduled = 2
	taskExecuted  = 3
	taskCancelled = 4
)

func nowMillis() int64 { return time.Now().UnixMilli() }

// TimerTask is the Go analogue of Java's abstract TimerTask. The abstract
// execute() is modelled by the exec func field (supplied by the caller).
type TimerTask struct {
	mu                sync.Mutex
	state             int
	period            int64
	nextExecutionTime int64
	exec              func()
	name              string
}

func newTimerTask(name string, exec func()) *TimerTask {
	return &TimerTask{state: taskNew, period: 0, exec: exec, name: name}
}

func newPeriodicTimerTask(name string, period int64, exec func()) *TimerTask {
	if period < 0 {
		panic("Period can't be negative")
	}
	return &TimerTask{state: taskNew, period: period, exec: exec, name: name}
}

func (t *TimerTask) cancel() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	ret := t.state == taskScheduled
	t.state = taskCancelled
	return ret
}

func (t *TimerTask) execute() {
	if t.exec != nil {
		t.exec()
	}
}

// compareTo: a task is "less" if it is scheduled earlier.
func (t *TimerTask) compareTo(other *TimerTask) int {
	if other == t {
		return 0
	}
	return int(t.getNextExecutionTime() - other.getNextExecutionTime())
}

func (t *TimerTask) setState(state int) {
	t.mu.Lock()
	t.state = state
	t.mu.Unlock()
}

func (t *TimerTask) getState() int {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.state
}

func (t *TimerTask) isPeriodic() bool { return t.period > 0 }

func (t *TimerTask) getNextExecutionTime() int64 {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.nextExecutionTime
}

func (t *TimerTask) setNextExecutionTime(v int64) {
	t.mu.Lock()
	t.nextExecutionTime = v
	t.mu.Unlock()
}

func (t *TimerTask) getPeriod() int64 { return t.period }

// taskHeap is a min-heap of *TimerTask ordered by next execution time.
type taskHeap []*TimerTask

func (h taskHeap) Len() int           { return len(h) }
func (h taskHeap) Less(i, j int) bool { return h[i].getNextExecutionTime() < h[j].getNextExecutionTime() }
func (h taskHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *taskHeap) Push(x any)        { *h = append(*h, x.(*TimerTask)) }
func (h *taskHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	old[n-1] = nil
	*h = old[:n-1]
	return x
}

// TimerQueue holds multiple timers in a priority queue and executes them on
// schedule. Java's wait/notify monitor is modelled with sync.Cond; the worker
// thread/Runnable is a goroutine.
type TimerQueue struct {
	mu     sync.Mutex
	cond   *sync.Cond
	heap   taskHeap
	stop   bool
	stopCh chan struct{}
}

func NewTimerQueue() *TimerQueue {
	tq := &TimerQueue{stopCh: make(chan struct{})}
	tq.cond = sync.NewCond(&tq.mu)
	return tq
}

// schedule runs the task immediately.
func (tq *TimerQueue) schedule(t *TimerTask) { tq.scheduleDelay(t, 0) }

// scheduleDelay runs the task after delay ms (renamed Java overload schedule(t,delay)).
func (tq *TimerQueue) scheduleDelay(t *TimerTask, delay int64) {
	if t == nil {
		panic("Can't schedule a null TimerTask")
	}
	if delay < 0 {
		delay = 0
	}
	t.setNextExecutionTime(nowMillis() + delay)
	tq.putJob(t)
}

func (tq *TimerQueue) putJob(task *TimerTask) {
	tq.mu.Lock()
	heap.Push(&tq.heap, task)
	task.setState(taskScheduled)
	tq.cond.Broadcast()
	tq.mu.Unlock()
}

// getJob blocks until a live task is available, or returns nil if stopped.
func (tq *TimerQueue) getJob() *TimerTask {
	tq.mu.Lock()
	defer tq.mu.Unlock()
	for {
		for len(tq.heap) == 0 && !tq.stop {
			tq.cond.Wait()
		}
		if tq.stop {
			return nil
		}
		task := heap.Pop(&tq.heap).(*TimerTask)
		switch task.getState() {
		case taskCancelled, taskExecuted:
			continue // skip dead task and fetch the next one
		case taskNew, taskScheduled:
			return task
		default:
			panic("TimerTask has an illegal state")
		}
	}
}

func (tq *TimerQueue) executeTask() func() { return tq.timerTaskLoop }

func (tq *TimerQueue) clear() {
	tq.mu.Lock()
	tq.heap = tq.heap[:0]
	tq.mu.Unlock()
}

// Stop terminates the worker loop and wakes any waiters.
func (tq *TimerQueue) Stop() {
	tq.mu.Lock()
	if !tq.stop {
		tq.stop = true
		close(tq.stopCh)
		tq.cond.Broadcast()
	}
	tq.mu.Unlock()
}

func (tq *TimerQueue) isStopped() bool {
	tq.mu.Lock()
	defer tq.mu.Unlock()
	return tq.stop
}

// sleepOrStop waits up to d, returning true if the queue was stopped meanwhile.
func (tq *TimerQueue) sleepOrStop(d time.Duration) bool {
	if d <= 0 {
		d = time.Millisecond
	}
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-timer.C:
		return tq.isStopped()
	case <-tq.stopCh:
		return true
	}
}

// timerTaskLoop is the worker: get next job, wait until due, then execute.
func (tq *TimerQueue) timerTaskLoop() {
	defer tq.clear()
	for {
		task := tq.getJob()
		if task == nil {
			return // stopped
		}
		now := nowMillis()
		executionTime := task.getNextExecutionTime()
		timeToWait := executionTime - now
		if timeToWait > 0 {
			// Not yet time to run the first one: put it back and wait.
			tq.putJob(task)
			if tq.sleepOrStop(time.Duration(timeToWait) * time.Millisecond) {
				return
			}
		} else {
			if task.isPeriodic() {
				// Reschedule with the new time.
				task.setNextExecutionTime(executionTime + task.getPeriod())
				tq.putJob(task)
			} else {
				// One-shot task already removed from the heap; mark executed.
				task.setState(taskExecuted)
			}
			task.execute()
		}
	}
}

func main() {
	fmt.Println("TimerQueue demo")

	tq := NewTimerQueue()
	var wg sync.WaitGroup
	var mu sync.Mutex
	var order []string

	scheduleOne := func(name string, delay int64) {
		wg.Add(1)
		t := newTimerTask(name, func() {
			mu.Lock()
			order = append(order, name)
			mu.Unlock()
			fmt.Printf("Executed %s\n", name)
			wg.Done()
		})
		tq.scheduleDelay(t, delay)
	}

	loop := tq.executeTask()
	go loop()

	scheduleOne("task-30ms", 30)
	scheduleOne("task-10ms", 10)
	scheduleOne("task-20ms", 20)

	wg.Wait()
	tq.Stop()

	fmt.Println("Execution order:", order)
}

File: Coding/python/timer_queue.py
"""Timer implemented with a priority queue of scheduled tasks.

Idiomatic Python 3 port of java/TimerQueue.java (original package code.ds).
The Java monitor (wait/notify) + worker Runnable thread + PriorityQueue are
modelled with ``threading.Condition`` + ``threading.Thread`` + ``heapq``.

The implementation is kept DETERMINISTIC and TERMINATING: the worker thread is a
daemon, ``stop()`` wakes it via the condition and joins it, and all demo waits
use timeouts so nothing can hang.
"""

import heapq
import threading
import time

# TimerTask states.
NEW = 1        # before the first execution
SCHEDULED = 2  # after first execution if repeating
EXECUTED = 3   # after first execution if not repeating
CANCELLED = 4  # when cancelled


def _now_millis():
    return time.monotonic() * 1000.0


class TimerTask:
    """Abstract analogue of Java's ``TimerTask``; subclasses implement execute()."""

    _seq_counter = 0
    _seq_lock = threading.Lock()

    def __init__(self, period=0):
        if period < 0:
            raise ValueError("Period can't be negative")
        self._lock = threading.Lock()
        self._state = NEW
        self._period = period  # constant; need not be locked
        self._next_execution_time = 0.0
        # Monotonic sequence number used only to break heap ties deterministically.
        with TimerTask._seq_lock:
            self._seq = TimerTask._seq_counter
            TimerTask._seq_counter += 1

    def cancel(self):
        """Cancel the next execution; returns True if a scheduled run is prevented."""
        with self._lock:
            ret = self._state == SCHEDULED
            self._state = CANCELLED
            return ret

    def execute(self):
        """The task to run, implemented in subclasses."""
        raise NotImplementedError

    # A task is "less than" another if it is scheduled earlier (Comparable).
    def __lt__(self, other):
        a = self.get_next_execution_time()
        b = other.get_next_execution_time()
        if a != b:
            return a < b
        return self._seq < other._seq

    def set_state(self, state):
        with self._lock:
            self._state = state

    def get_state(self):
        with self._lock:
            return self._state

    def is_periodic(self):
        return self._period > 0

    def get_next_execution_time(self):
        with self._lock:
            return self._next_execution_time

    def set_next_execution_time(self, t):
        with self._lock:
            self._next_execution_time = t

    def get_period(self):
        return self._period


class FuncTimerTask(TimerTask):
    """Concrete TimerTask whose execute() invokes a supplied callable."""

    def __init__(self, name, func, period=0):
        super().__init__(period)
        self.name = name
        self._func = func

    def execute(self):
        if self._func is not None:
            self._func()


class TimerQueue:
    def __init__(self):
        self._cond = threading.Condition()  # guards the heap and the stop flag
        self._heap = []                      # min-heap of TimerTask by next time
        self._stop = False
        self._worker = None

    def start(self):
        """Launch the worker thread that executes tasks on schedule."""
        if self._worker is None:
            self._worker = threading.Thread(target=self._timer_task_loop, daemon=True)
            self._worker.start()

    # schedule(t) and schedule(t, delay) merged via a default delay of 0.
    def schedule(self, t, delay=0):
        if t is None:
            raise ValueError("Can't schedule a null TimerTask")
        if delay < 0:
            delay = 0
        t.set_next_execution_time(_now_millis() + delay)
        self._put_job(t)

    def _put_job(self, task):
        with self._cond:
            heapq.heappush(self._heap, task)
            task.set_state(SCHEDULED)
            self._cond.notify_all()

    def _get_job(self):
        """Block until a live task is available; return None once stopped."""
        with self._cond:
            while True:
                while not self._heap and not self._stop:
                    self._cond.wait()
                if self._stop:
                    return None
                task = heapq.heappop(self._heap)
                st = task.get_state()
                if st in (CANCELLED, EXECUTED):
                    # Don't hold onto a dead task; fetch the next one.
                    continue
                if st in (NEW, SCHEDULED):
                    return task
                raise RuntimeError("TimerTask has an illegal state")

    def clear(self):
        with self._cond:
            self._heap.clear()

    def stop(self):
        """Terminate the worker loop and wake any waiters, then join."""
        with self._cond:
            if not self._stop:
                self._stop = True
                self._cond.notify_all()
        if self._worker is not None:
            self._worker.join(timeout=2.0)
            self._worker = None

    def _wait_or_stop(self, millis):
        """Wait up to `millis` ms (or until stopped/notified). Return True if stopped."""
        if millis <= 0:
            millis = 1
        with self._cond:
            if self._stop:
                return True
            self._cond.wait(timeout=millis / 1000.0)
            return self._stop

    def _timer_task_loop(self):
        """Worker: get the next job, wait until it is due, then execute it."""
        try:
            while True:
                task = self._get_job()
                if task is None:
                    return  # stopped
                now = _now_millis()
                execution_time = task.get_next_execution_time()
                time_to_wait = execution_time - now
                if time_to_wait > 0:
                    # A job was extracted but it is not yet time to run it:
                    # reschedule (put it back) and wait.
                    self._put_job(task)
                    if self._wait_or_stop(time_to_wait):
                        return
                else:
                    if task.is_periodic():
                        # Reschedule with the new time.
                        task.set_next_execution_time(execution_time + task.get_period())
                        self._put_job(task)
                    else:
                        # The one-shot task is already removed from the heap.
                        task.set_state(EXECUTED)
                    # Run it!
                    task.execute()
        finally:
            self.clear()


if __name__ == "__main__":
    print("TimerQueue demo")

    tq = TimerQueue()
    tq.start()

    results = []
    rlock = threading.Lock()
    counter = {"n": 0}
    done = threading.Event()
    total = 3

    def make_run(name):
        def run():
            with rlock:
                results.append(name)
                counter["n"] += 1
                if counter["n"] >= total:
                    done.set()
        return run

    # Three one-shot tasks due at 10/20/30 ms -> executed in that (deterministic) order.
    tq.schedule(FuncTimerTask("task-10ms", make_run("task-10ms")), 10)
    tq.schedule(FuncTimerTask("task-30ms", make_run("task-30ms")), 30)
    tq.schedule(FuncTimerTask("task-20ms", make_run("task-20ms")), 20)

    finished = done.wait(timeout=5.0)  # bounded wait so the demo can never hang
    tq.stop()

    print("all tasks executed:", finished)
    print("execution order:", results)
